File: app/connector/webhook.py
import logging
from typing import Any
from typing import Literal

import httpx
from pydantic import BaseModel

from .enum import ConnectorType
from app.connector.base import BaseConnector

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


class WebhookConfig(BaseModel):
    """Configuration for Webhook connector."""

    url: str
    method: str
    headers: dict[str, str] = {}
    body: dict[str, Any] = {}


class WebhookWorkflowStep(BaseModel):
    """Definition of a Webhook step in a workflow."""

    type: Literal[ConnectorType.WEBHOOK] = ConnectorType.WEBHOOK
    name: str
    config: WebhookConfig


class WebhookResponse(BaseModel):
    """Output model for Webhook connector."""

    type: Literal[ConnectorType.WEBHOOK] = ConnectorType.WEBHOOK
    status_code: int
    response_data: Any
    url: str
    method: str


class WebhookConnector(BaseConnector):
    """Connector that makes HTTP requests."""

    def __init__(self):
        super().__init__(ConnectorType.WEBHOOK)

    async def execute(
        self, step: WebhookWorkflowStep, context: dict[str, Any]
    ) -> WebhookResponse:
        """
        Make HTTP request to webhook URL.

        Args:
            step (WebhookWorkflowStep): The step configuration.
            context (dict[str, Any]): The execution context.

        Returns:
            WebhookResponse: The response from the webhook.

        Raises:
            ValueError: If an unsupported HTTP method is used.
        """
        url = step.config.url
        method = step.config.method.upper()
        headers = step.config.headers
        body = step.config.body

        # Replace placeholders in body with context data
        if isinstance(body, dict):
            body = self._replace_placeholders(body, context)

        logger.info(f"Making {method} request to {url}")

        async with httpx.AsyncClient() as client:
            if method == "GET":
                response = await client.get(url, headers=headers)
            elif method == "POST":
                response = await client.post(url, json=body, headers=headers)
            elif method == "PUT":
                response = await client.put(url, json=body, headers=headers)
            elif method == "DELETE":
                response = await client.delete(url, headers=headers)
            else:
                raise ValueError(f"Unsupported HTTP method: {method}")

        response_data = (
            response.json()
            if response.headers.get("content-type", "").startswith("application/json")
            else response.text
        )
        return WebhookResponse(
            status_code=response.status_code,
            response_data=response_data,
            url=url,
            method=method,
        )

    def _replace_placeholders(self, data: Any, context: dict[str, Any]) -> Any:
        """
        Replace placeholders in data with context values.
        Supports dictionary, list, and string replacement (e.g., "${key}").

        Args:
            data (Any): The data containing placeholders.
            context (dict[str, Any]): The context values.

        Returns:
            Any: The data with placeholders replaced.
        """
        if isinstance(data, dict):
            return {k: self._replace_placeholders(v, context) for k, v in data.items()}
        elif isinstance(data, list):
            return [self._replace_placeholders(item, context) for item in data]
        elif isinstance(data, str) and data.startswith("${") and data.endswith("}"):
            # Simple placeholder replacement: ${key} -> context[key]
            key = data[2:-1]
            return context.get(key, data)
        return data

File: rust-app/src/connector/webhook.rs
use async_trait::async_trait;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use tracing::info;

use crate::connector::base::Connector;
use crate::schemas::workflow::{WorkflowStep, WorkflowStepResponse};

/// Configuration for the Webhook connector.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WebhookConfig {
    pub url: String,
    pub method: String,
    #[serde(default)]
    pub headers: HashMap<String, String>,
    #[serde(default)]
    pub body: serde_json::Value,
}

/// Output model for the Webhook connector.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WebhookResponse {
    pub status_code: u16,
    pub response_data: serde_json::Value,
    pub url: String,
    pub method: String,
}

/// Connector that makes HTTP requests.
pub struct WebhookConnector;

impl WebhookConnector {
    pub fn new() -> Self {
        Self
    }
}

impl Default for WebhookConnector {
    fn default() -> Self {
        Self::new()
    }
}

#[async_trait]
impl Connector for WebhookConnector {
    async fn execute(
        &self,
        step: &WorkflowStep,
        context: &HashMap<String, serde_json::Value>,
    ) -> Result<WorkflowStepResponse, Box<dyn std::error::Error + Send + Sync>> {
        match step {
            WorkflowStep::Webhook { config, .. } => {
                let url = &config.url;
                let method = config.method.to_uppercase();
                let headers = &config.headers;
                let body = Self::replace_placeholders(&config.body, context);

                info!("Making {} request to {}", method, url);

                let client = reqwest::Client::new();
                let mut request_builder = match method.as_str() {
                    "GET" => client.get(url),
                    "POST" => client.post(url).json(&body),
                    "PUT" => client.put(url).json(&body),
                    "DELETE" => client.delete(url),
                    _ => return Err(format!("Unsupported HTTP method: {}", method).into()),
                };

                for (key, value) in headers {
                    request_builder = request_builder.header(key.as_str(), value.as_str());
                }

                let response = request_builder.send().await?;
                let status_code = response.status().as_u16();
                let content_type = response
                    .headers()
                    .get("content-type")
                    .and_then(|v| v.to_str().ok())
                    .unwrap_or("")
                    .to_string();

                let response_data = if content_type.starts_with("application/json") {
                    response.json::<serde_json::Value>().await?
                } else {
                    serde_json::Value::String(response.text().await?)
                };

                Ok(WorkflowStepResponse::Webhook(WebhookResponse {
                    status_code,
                    response_data,
                    url: url.clone(),
                    method,
                }))
            }
            _ => Err("WebhookConnector received non-webhook step".into()),
        }
    }
}

impl WebhookConnector {
    /// Replace placeholders in data with context values.
    /// Supports ${key} syntax for simple string replacement.
    fn replace_placeholders(
        data: &serde_json::Value,
        context: &HashMap<String, serde_json::Value>,
    ) -> serde_json::Value {
        match data {
            serde_json::Value::Object(map) => {
                let mut new_map = serde_json::Map::new();
                for (k, v) in map {
                    new_map.insert(k.clone(), Self::replace_placeholders(v, context));
                }
                serde_json::Value::Object(new_map)
            }
            serde_json::Value::Array(arr) => {
                serde_json::Value::Array(
                    arr.iter()
                        .map(|item| Self::replace_placeholders(item, context))
                        .collect(),
                )
            }
            serde_json::Value::String(s) if s.starts_with("${") && s.ends_with('}') => {
                let key = &s[2..s.len() - 1];
                // Look up in context by key
                if let Some(val) = context.get(key) {
                    val.clone()
                } else {
                    data.clone()
                }
            }
            _ => data.clone(),
        }
    }
}
